pkg/manager/manager.go

```diff
@@ -0,0 +1,310 @@
+			return e
+		}
+		return wErr
+	}
+	m.BundleFileName = bundleName
+
+	err = m.waitNodeBundles()
+	if err != nil {
+		if e := m.harvester.SetSupportBundleError(m.BundleName, StateError, err.Error()); e != nil {
+			return e
+		}
+		return err
+	}
+
+	err = m.harvester.UpdateSupportBundleStatus2(m.BundleName, StateReady, m.BundleFileName, m.bundleFileSize)
+	if err != nil {
+		return err
+	}
+
+	select {}
+}
+
+func (m *SupportBundleManager) waitNodeBundles() error {
+	m.ch = make(chan struct{})
+
+	nodes, err := m.getHarvesterNodes()
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("expected bundles from nodes: %+v", nodes)
+	m.expectedNodeCount = len(nodes)
+
+	// create a http server to receive node bundles
+	s := HttpServer{context: m.context}
+	go s.Run(m)
+
+	// create a daemonset to collect node bundles and push back
+	agents := &AgentDaemonSet{sbm: m}
+	err = agents.Create(m.ImageName, fmt.Sprintf("http://%s:8080", m.ManagerPodIP))
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("wating node bundles, (timeout: %s)", m.WaitTimeout)
+	select {
+	case <-m.ch:
+		logrus.Debug("all node bundles are received.")
+		err := m.compressBundle()
+		if err != nil {
+			return errors.Wrap(err, "fail to compress bundle")
+		}
+
+		size, err := m.getBundlefilesize()
+		if err != nil {
+			return errors.Wrap(err, "fail to get bundle file size")
+		}
+		m.bundleFileSize = size
+
+		// Clean up when everything is fine, leave ds there for debugging.
+		// The ds will be garbage-collected when manager pod is gone
+		return agents.Cleanup()
+	case <-time.After(m.WaitTimeout):
+		return errors.New("fail to wait for all node bundles")
+	}
+}
+
+func (m *SupportBundleManager) getBundle(w http.ResponseWriter, req *http.Request) {
+	bundleFile := m.getBundlefile()
+	f, err := os.Open(bundleFile)
+	if err != nil {
+		e := errors.Wrap(err, "fail to open bundle file")
+		logrus.Error(e)
+		utils.HttpResponseError(w, http.StatusNotFound, e)
+		return
+	}
+	defer f.Close()
+
+	fstat, err := f.Stat()
+	if err != nil {
+		e := errors.Wrap(err, "fail to stat bundle file")
+		logrus.Error(e)
+		utils.HttpResponseError(w, http.StatusNotFound, e)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+	w.Header().Set("Content-Length", strconv.FormatInt(fstat.Size(), 10))
+	w.Header().Set("Content-Disposition", "attachment; filename="+filepath.Base(bundleFile))
+	if _, err := io.Copy(w, f); err != nil {
+		utils.HttpResponseError(w, http.StatusInternalServerError, err)
+		return
+	}
+}
+
+func (m *SupportBundleManager) createNodeBundle(w http.ResponseWriter, req *http.Request) {
+	node := mux.Vars(req)["nodeName"]
+	if node == "" {
+		return
+	}
+
+	logrus.Debugf("handle create node bundle for %s", node)
+	nodesDir := filepath.Join(m.getWorkingDir(), "nodes")
+	err := os.MkdirAll(nodesDir, os.FileMode(0775))
+	if err != nil {
+		logrus.Errorf("fail to create directory %s: %s", nodesDir, err)
+		return
```

Comment:
Is it expected that when an error happens we returns a statusOK

Reply:
[Fixed](d419fd6). Thanks!